src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/Extensions/ContextQuery/SyntaxTreeExtensions.cs

```diff
@@ -3017,5 +3017,98 @@ public static bool IsFunctionPointerCallingConventionContext(this SyntaxTree syn
+        /// </summary>
+        public static bool IsBaseEnumContext(this SyntaxTree syntaxTree, SyntaxToken targetToken)
+        {
+            // We don't care about qualified names, generics or whatever else, because in base list of enum can only contain one item of byte, short etc.
+            return targetToken.GetAncestor<BaseListSyntax>() is { Parent: EnumDeclarationSyntax };
+        }
+
+        /// <summary>
+        /// Tells if in a position of a base list, where interface name is a valid completion.
+        /// For instatnce, this flag will be <see langword="true" /> for
+        /// <code>class C : $$</code>
+        /// or
+        /// <code>interface I : $$</code>
+        /// </summary>
+        public static bool IsBaseInterfaceContext(this SyntaxTree syntaxTree, SyntaxToken targetToken)
+        {
+            // Case with non-written base type, e.g. "class C : $$" or "class C : Base, $$"
+            if (targetToken.Kind() is SyntaxKind.ColonToken or SyntaxKind.CommaToken &&
+                targetToken.Parent is BaseListSyntax { Parent: not EnumDeclarationSyntax })
+            {
+                return true;
+            }
+
+            return IsValidBaseTypeSyntaxQuilifiedNameToken(targetToken);
+        }
+
+        /// <summary>
+        /// Tells if in a position of a base list, where record name is a valid completion.
+        /// For instatnce, this flag will be <see langword="true" /> for
+        /// <code>record R : $$</code>
+        /// but not for
+        /// <code>record R : A, $$</code>
+        /// </summary>
+        public static bool IsBaseRecordContext(this SyntaxTree syntaxTree, SyntaxToken targetToken)
+        {
+            var possibleBaseListSyntax = targetToken.GetAncestor<BaseListSyntax>();
+            var possibleBaseTypeSyntax = targetToken.GetAncestor<BaseTypeSyntax>();
+
+            if (possibleBaseListSyntax?.Parent is not RecordDeclarationSyntax)
+                return false;
+
+            // We are in a non-written base type, e.g. "record R : $$"
+            if (possibleBaseTypeSyntax is null)
+                return targetToken.IsKind(SyntaxKind.ColonToken);
+
+            // Current base type must be the first in the base list, because record inheritance is valid only at the first position in the base list
+            if (!possibleBaseListSyntax.Types[0].Equals(possibleBaseTypeSyntax))
+                return false;
+
+            return IsValidBaseTypeSyntaxQuilifiedNameToken(targetToken);
+        }
+
+        /// <summary>
+        /// Checks if the token is in a valid position inside a BaseTypeSyntax including qualified names of any depth
+        /// </summary>
+        private static bool IsValidBaseTypeSyntaxQuilifiedNameToken(SyntaxToken token)
+        {
+            var parent = token.Parent;
+
+            while (parent is QualifiedNameSyntax or AliasQualifiedNameSyntax)
+                parent = parent.Parent;
+
+            return parent is BaseTypeSyntax;
```

Comment:
this doesn't seem like it will work for something like `class X : X$$<int>`.  It would say we're not in a valid base-type syntax case here afaict